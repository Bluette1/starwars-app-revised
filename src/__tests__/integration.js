const { createTestClient } = require('apollo-server-testing');
const gql = require('graphql-tag');
const { constructTestServer } = require('./__utils');

// the mocked REST API data
const { mockPersonResponse, mockPerson } = require('../datasources/__tests__/person');

const PEOPLE_LIST_QUERY = gql`
  query people($page: Int) {
    people(page: $page) {
      name
      height
      gender
      homeworld
    }
  }
`;

const GET_PERSON_BY_NAME = gql`
  query person($name: String) {
    person(name: $name) {
      name
      height
      gender
      homeworld
    }
  }
`;

const GET_PERSON_BY_ID = gql`
  query person($id: Int) {
    person(id: $id) {
      name
      height
      gender
      homeworld
    }
  }
`;

describe('Queries', () => {
  it('fetches list of people', async () => {
    // create an instance of ApolloServer that mocks out context, while reusing
    // existing dataSources, resolvers, and typeDefs.
    // This function returns the server instance as well as our dataSource
    // instances, so we can overwrite the underlying fetchers
    const { server, person } = constructTestServer({
      context: () => ({ userId: 1 }),
    });

    // mock the datasources' underlying fetch methods, whether that's a REST
    // lookup in the RESTDataSource or the store query in the Sequelize datasource
    person.get = jest.fn(() => ({ results: [mockPersonResponse] }));

    // use our test server as input to the createTestClient fn
    // This will give us an interface, similar to apolloClient.query
    // to run queries against our instance of ApolloServer
    const { query } = createTestClient(server);
    const res = await query({ query: PEOPLE_LIST_QUERY });
    expect(res).toMatchSnapshot();
    expect(res.data.people).toEqual([mockPerson]);
  });

  it('fetches single person by id', async () => {
    const { server, person } = constructTestServer({
      context: () => ({ userId: 1 }),
    });

    person.get = jest.fn(() => mockPersonResponse);

    const { query } = createTestClient(server);
    const res = await query({ query: GET_PERSON_BY_ID, variables: { id: 1 } });
    expect(res).toMatchSnapshot();
    expect(res.data.person).toEqual(mockPerson);
  });

  it('fetches single person by name', async () => {
    const { server, person } = constructTestServer({
      context: () => ({ userId: 1 }),
    });

    person.get = jest.fn(() => ({ results: [mockPersonResponse] }));

    const { query } = createTestClient(server);
    const res = await query({
      query: GET_PERSON_BY_NAME,
      variables: { name: 'Luke Skywalker' },
    });
    expect(res).toMatchSnapshot();
    expect(res.data.person).toEqual(mockPerson);
  });
});
