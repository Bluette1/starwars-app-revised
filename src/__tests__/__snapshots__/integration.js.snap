// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Queries fetches list of people 1`] = `
Object {
  "data": Object {
    "people": Array [
      Object {
        "gender": "male",
        "height": "172",
        "homeworld": "http://swapi.dev/api/planets/1/",
        "name": "Luke Skywalker",
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Queries fetches single person by id 1`] = `
Object {
  "data": Object {
    "person": Object {
      "gender": "male",
      "height": "172",
      "homeworld": "http://swapi.dev/api/planets/1/",
      "name": "Luke Skywalker",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Queries fetches single person by name 1`] = `
Object {
  "data": Object {
    "person": Object {
      "gender": "male",
      "height": "172",
      "homeworld": "http://swapi.dev/api/planets/1/",
      "name": "Luke Skywalker",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
